refactor: PostListing and SearchResults (#164)
MovingBlocks/movingblocks.github.com: _posts/2021-02-07-teraspotlight-lod-chunks.md

---
layout: post

> title: "TeraSpotlight: Extrem View Distances with LOD Chunks"

Author: ```suggestion
title: "TeraSpotlight: Extreme View Distances with LOD Chunks"
```

description: >
  The idea of extreme view distances with LOD chunks is nearly as old as this project, and now it's finally a dream come true.
author: Skaldarnar

> header-img: "/img/terasaturday/9/banner.jpg"

Author: is that intentionally in the terasaturday folder

Reviewer: forgot to push latest changes - it now uses it's own banner image from the correct folder.

---

_Please enjoy this special spotlight post about a new feature we are excited about. We hope you will like it as much as we do!_

## 🤩 Spotlight

Thanks to _@4D Enthusiast_ a long-time dream has finally come true. 

> You can experience the worlds of Terasology with [**extrem view and render distances**](https://github.com/MovingBlocks/Terasology/pull/4452) now!

Author: ```suggestion
You can experience the worlds of Terasology with [**extreme view and render distances**](https://github.com/MovingBlocks/Terasology/pull/4452) now!
```

    </div>
</div>

> For each level of detail (with the number controlled by a slider, 1 to 10), a set of chunks is generated with half the resolution, but twice the size.
> There are also some LOD chunks around the edge of the actual loaded region that are about the same size as real chunks, in order to ensure that everything lines up neatly if the edge of the loaded region does not align to the double-scale chunk grid. 
> This means that the amount of detail visible at each level of detail is approximately the same (because of perspective), but size of the visible area grows exponentially with the number of LOD levels. 
> I set the maximum LOD level to 10 not because adding more would be too computationally expensive for anyone to run, but because the view area is already so large that levels 9 and 10 look practically identical anyway. 
> As an example, with view distance Near (9 * 7 * 9), it would take 16 LODs to have an area the size of the earth visible, and 22 LODs for the moon to be directly visible in the sky (although a block-based moon couldn't actually move).

So, the available levels of details should be sufficient 👍 

> In addition to scaling down the chunk resolution the visible details are also limited to _world facets_ that support the scaling.

Author: ```suggestion
In addition to scaling down the chunk resolution, the visible details are also limited to _world facets_ that support the scaling.
```

> This is currently only implemented for the basic terrain generation, but we are looking forward to extending this to features like trees.

Author: Maybe mention `CoreSampleGameplay` here so people know what to start to check it out?

Reviewer: It also works for JS I think :thinking:

>
> In order to avoid issues with ambient occlusion and face culling, each LOD chunk is actually generated slightly larger than the area it needs to cover, so that there's overlap between the edges of adjacent chunks' meshes. 
> This also helps to cover up some of the gaps caused by differences between chunks of different scales.
> I also tried at one point having all the faces on the outside of each chunk visible in order to avoid gaps, but that made it much less efficient because of all of the underground chunks that would become visible.

<div class="row">
    <div class="col s12 m6 l6">
        <img class="img-margin fwidth materialboxed z-depth-3" src="{{ site.baseurl }}/img/2021-02-07-teraspotlight-lod-chunks/Terasology-210201224636-1920.jpg" alt="{{ image.title }}">
    </div>
    <div class="col s12 m6 l6">
        <img class="img-margin fwidth materialboxed z-depth-3" src="{{ site.baseurl }}/img/2021-02-07-teraspotlight-lod-chunks/Terasology-210201231714-1920.jpg" alt="{{ image.title }}">
    </div>
</div>

> Another issue I had to fix was that the far clipping plane was sometimes too close. 

> > There's a tradeoof in rendering between the accuracy with which it can be determined which of two things is closer (and therefore should be visible in front) and the maximum distance at which you can see something. 

Author: ```suggestion
> There's a trade-off in rendering between the accuracy with which it can be determined which of two things is closer (and therefore should be visible in front) and the maximum distance at which you can see something. 
```
